src/Time/Slab.hpp

```diff
@@ -0,0 +1,137 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+/// \file
+/// Defines class Slab
+
+#pragma once
+
+#include <functional>
+#include <limits>
+
+#include "ErrorHandling/Assert.hpp"
+
+class Time;
+class TimeDelta;
+
+namespace PUP {
+class er;
+}
+
+/// A chunk of time.  Every element must reach slab boundaries
+/// exactly, no matter how it actually takes time steps to get there.
+/// The simulation can only be assumed to have global data available
+/// at slab boundaries.
+class Slab {
+ public:
+  /// Default constructor gives an invalid Slab.
+  Slab() noexcept
+      : start_(std::numeric_limits<double>::signaling_NaN()),
+        end_(std::numeric_limits<double>::signaling_NaN()) {}
+
+  /// Construct a slab running between two times (exactly).
+  Slab(double start, double end) noexcept : start_(start), end_(end) {
+    ASSERT(start_ < end_, "Backwards Slab");
+  }
+
+  /// Construct a slab with a given start time and duration.  The
+  /// actual duration may differ by roundoff from the supplied value.
+  static Slab new_with_duration(double start, double duration) noexcept {
```

Comment:
Okay, so for the static function I felt the new was not needed as it obviously had to make a new object.  I have a slight concern about the member variable with_* being misused by someone thinking it modifies the object instead of creating a new one, hence my suggestion of create_with_*, but I understand your point about this naming being the opposite of our factory-like convention.  Since they both create a new object and have different interfaces, I would be okay with them sharing the same name.  Anyone else have an opinion?